win32ss/gdi/dib/dib24bpp.c

```diff
@@ -96,65 +177,158 @@ DIB_24BPP_BitBltSrcCopy(PBLTINFO BltInfo)
-          SourceBits += 1;
+          if (bLeftToRight)
+          {
+              SourceBits -= 1;
+          }
+          else
+          {
+              SourceBits += 1;
+          }
           DestBits += 3;
         }
 
-        SourceLine += BltInfo->SourceSurface->lDelta;
+        if (bTopToBottom)
+        {
+          SourceLine -= BltInfo->SourceSurface->lDelta;
+        }
+        else
+        {
+          SourceLine += BltInfo->SourceSurface->lDelta;
+        }
         DestLine += BltInfo->DestSurface->lDelta;
       }
       break;
 
     case BMF_16BPP:
+      DPRINT("16BPP Case Selected with DestRect Width of '%d'.\n",
+        BltInfo->DestRect.right - BltInfo->DestRect.left);
+
+      /* This sets SourceBits_16BPP to the top line */
       SourceBits_16BPP = (PWORD)((PBYTE)BltInfo->SourceSurface->pvScan0 + (BltInfo->SourcePoint.y * BltInfo->SourceSurface->lDelta) + 2 * BltInfo->SourcePoint.x);
 
+      if (bTopToBottom)
+      {
+        /* This sets SourceBits_16BPP to the bottom line */
+        SourceBits_16BPP += (BltInfo->DestRect.bottom - BltInfo->DestRect.top - 1) * BltInfo->SourceSurface->lDelta;
+      }
+
       for (j=BltInfo->DestRect.top; j<BltInfo->DestRect.bottom; j++)
       {
         SourceLine_16BPP = SourceBits_16BPP;
         DestLine = DestBits;
 
+        if (bLeftToRight)
+        {
+          /* This sets the SourceLine_16BPP to the rightmost pixel */
+          SourceLine_16BPP += (BltInfo->DestRect.right - BltInfo->DestRect.left - 1);
+        }
+
         for (i=BltInfo->DestRect.left; i<BltInfo->DestRect.right; i++)
         {
           xColor = XLATEOBJ_iXlate(BltInfo->XlateSourceToDest, *SourceLine_16BPP);
           *DestLine++ = xColor & 0xff;
           *(PWORD)DestLine = (WORD)(xColor >> 8);
           DestLine += 2;
-          SourceLine_16BPP++;
+          if (bLeftToRight)
+          {
+            SourceLine_16BPP--;
+          }
+          else
+          {
+            SourceLine_16BPP++;
+          }
+        }
+        if (bTopToBottom)
+        {
+          SourceBits_16BPP = (PWORD)((PBYTE)SourceBits_16BPP - BltInfo->SourceSurface->lDelta);
+        }
+        else
+        {
+          SourceBits_16BPP = (PWORD)((PBYTE)SourceBits_16BPP + BltInfo->SourceSurface->lDelta);
         }
-
-        SourceBits_16BPP = (PWORD)((PBYTE)SourceBits_16BPP + BltInfo->SourceSurface->lDelta);
         DestBits += BltInfo->DestSurface->lDelta;
       }
       break;
 
     case BMF_24BPP:
-      if (NULL == BltInfo->XlateSourceToDest || 0 != (BltInfo->XlateSourceToDest->flXlate & XO_TRIVIAL))
+      DPRINT("24BPP Case Selected with DestRect Width of '%d'.\n",
+        BltInfo->DestRect.right - BltInfo->DestRect.left);
+
+      /* Check for no flips here because we are about to use RtlMoveMemory and it can only do increasing src & dst */
+      if ((NULL == BltInfo->XlateSourceToDest || 0 != (BltInfo->XlateSourceToDest->flXlate & XO_TRIVIAL)) &&
+        (!bTopToBottom && !bLeftToRight))
```

Comment:
Since you are changing this line, please fix inverse logic.
```suggestion
      if ((BltInfo->XlateSourceToDest == NULL ||
        (BltInfo->XlateSourceToDest->flXlate & XO_TRIVIAL) != 0) &&
        (!bTopToBottom && !bLeftToRight))
```